amp-bind: Fix embedding in FIF

choumx opened the pull request:
Partial for #9395.

- Adds support for retrieving element services in embed scopes
- Adds `services` to `ExtensionDef` -- now an extension knows which services it registers
- Minor changes to `installExtensionsInChildWindow`

choumx commented:
/to @dvoytenko

dvoytenko commented:
@choumx Thanks! Could we quickly discuss the overall plan? How would amp-bind work in FIE?

choumx commented:
@dvoytenko Sure! I split that part off into #9447. 

- `Bind#adoptEmbedWindow` creates a new Bind service and registers it to the embed window. It scans the embed window's document for bindings and has its own variable scope
- All Bind services (top-level, embeds) share a single Worker. I added window-level scoping to the Worker (see #9447 for details).

I'd love to discuss/work on:
- AMP Roots, since it's a bit weird that ampdoc-scope services are actually window-level in FIE. 
- Multiple elements per extension (#9143) to remove special-casing for Bind in this PR.

Thanks Dima.

dvoytenko commented on src/custom-element.js:
```diff
@@ -219,7 +219,11 @@ export function copyElementToChildWindow(parentWin, childWin, name) {
 export function upgradeElementInChildWindow(parentWin, childWin, name) {
   const toClass = getExtendedElements(parentWin)[name];
-  dev().assert(toClass, '%s is not stubbed yet', name);
+  // Some extensions unofficially register unstubbed elements, e.g. amp-bind.
```
Not sure I got this. Could you please explain?

choumx replied:
Extensions stub an element of the same name, but `amp-bind` registers the `amp-state` element which won't be stubbed. That's okay since `amp-state` is really just metadata, but we shouldn't `assert()` here for such elements. #9143 will make this distinction clearer.

dvoytenko commented on src/service/extensions-impl.js:
```diff
@@ -410,6 +410,12 @@ export class Extensions {
         // becomes necessary. This will require refactoring of extension
         // loader that can be resolved via the parent ampdoc.
 
+        // TODO(choumx): Similar to `adoptServicesForEmbed`, should there be
```
I don't think we can have a mapping like this. Maybe instead the extension itself can somehow mark itself as "adoptable to embed"? Or an extension could register an adopter?

choumx replied:
Good idea. Now an extension knows what services it registers.

dvoytenko commented on src/service/extensions-impl.js:
```diff
@@ -420,12 +426,15 @@ export class Extensions {
         // Notice that stubbing happens much sooner above
         // (see stubElementInChildWindow).
-        upgradeElementInChildWindow(topWin, childWin, extensionId);
+        Object.keys(extension.elements).forEach(element => {
```
E.g. here, why can't we also have `extension.services.forEach(service => if (isAdoptable(service) {service.adopt()}))`?

choumx replied:
This is just for upgrading stubs. I changed it to use the extension elements since it's more precise.

dvoytenko commented:
@choumx LGTM. This is actually a huge improvement :)
